src/decisionengine_modules/NERSC/sources/NerscSFApi.py

```diff
@@ -0,0 +1,146 @@
+        # ssiadmin4:/srv/secrets/per_host/hepcsvc03.fnal.gov/credentials/
+        pemfile_ucms = self.constraints.get("ucms_private_key")
+        pemfile_fife = self.constraints.get("fife_private_key")
+        client_id_ucms = self.constraints.get("ucms_client_id")
+        client_id_fife = self.constraints.get("fife_client_id")
+
+        rawfile = None
+        pemfile = None
+        client_id = None
+        if nersc_user == "uscms":
+            rawfile = rawfile_ucms
+            pemfile = pemfile_ucms
+            with open(client_id_ucms) as ifile:
+                client_id = ifile.read()
+                client_id = client_id.rstrip()
+        elif nersc_user == "fife":
+            rawfile = rawfile_fife
+            pemfile = pemfile_fife
+            with open(client_id_fife) as ifile:
+                client_id = ifile.read()
+                client_id = client_id.rstrip()
+        else:
+            self.logger.error("Unknown user, exiting")
+            return None
+
+        if not os.path.exists(rawfile):
+            self.logger.debug(f"{rawfile} does not exist. Need to generate")
+            renew_bool = True
+        else:
+            atoken = None
+            with open(rawfile) as afile:
+                atoken = afile.read()
+                atoken = atoken.rstrip()
+            # HK> If the access token is expired, the flow goes directly to except jwt.ExpiredSignatureError
+            try:
+                #                result = jwt.decode(atoken, options={"verify_signature": False})
+                jwt.decode(atoken, options={"verify_signature": False})
+                self.logger.debug("Not expired. Returning without generating a new access token")
+                return atoken  # This means the existing access token is not expired.
+
+            except jwt.ExpiredSignatureError:
+                self.logger.debug("Access Token expired")
+                renew_bool = True
+
+        if renew_bool:
+            certs = pem.parse_file(pemfile)
+            private_key = str(certs[0])
+            client = OAuth2Session(
+                client_id=client_id, client_secret=private_key, token_endpoint_auth_method="private_key_jwt"
+            )
+            client.register_client_auth_method(PrivateKeyJWT(token_url))
+            resp = client.fetch_token(token_url, grant_type="client_credentials")
+
+            newtoken = resp["access_token"]
+
+            with open(rawfile, "w") as myfile:
+                myfile.write(newtoken)
+            return newtoken
+
+    def get_headers2(self, access_token):
+        headers = {}
+        headers["content-type"] = "application/x-www-form-urlencoded"
+        headers["Authorization"] = "Bearer %s" % access_token
+        return headers
+
+    def requests_nersc(self, username):
+        my_access_token = self.check_accesstoken(username)
+        my_header = self.get_headers2(my_access_token)
+        p_url = "https://api.nersc.gov/api/v1.2/account/projects"
```

Comment:
This url and the one earlier could go on constants on top of the file, easy to update in case NERSC changes them.  Optional, I don't feel strongly about this, the URLs may stay the same for the lifetime of the API

Reply:
I updated the code to look
token_url = self.constraints.get("token_url")
with comments
# /etc/decisionengine/config.d/Nersc.jsonnet needs to also have                                                
# "token_url": "https://oidc.nersc.gov/c2id/token",